src/entities/nestedExit/getNestedExitCalls.ts

```diff
@@ -0,0 +1,193 @@
+import { Token } from '../token';
+import { NestedExitInput, NestedExitCall } from './types';
+import { NestedPool, NestedPoolState } from '../types';
+import { TokenAmount } from '../tokenAmount';
+import { Address } from '../../types';
+
+export const getNestedExitCalls = (
+    {
+        bptAmountIn,
+        chainId,
+        accountAddress,
+        tokenOut,
+        useNativeAssetAsWrappedAmountOut = false,
+        toInternalBalance = false,
+    }: NestedExitInput,
+    { pools }: NestedPoolState,
+): { bptAmountIn: TokenAmount; calls: NestedExitCall[] } => {
+    const isProportional = tokenOut === undefined;
+    let poolsTopDown: NestedPool[];
+    let calls: NestedExitCall[];
+
+    if (isProportional) {
+        /**
+         * Overall logic to build sequence of proportional exit calls:
+         * 1. Go from top pool to bottom filling out input amounts and output refs
+         * 2. Inputs will be bptAmountIn provided or output of the previous level
+         * 3. Output at bottom level is the amountsOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+
+        calls = getProportionalExitCalls(
+            poolsTopDown,
+            chainId,
+            useNativeAssetAsWrappedAmountOut,
+            accountAddress,
+            bptAmountIn,
+            toInternalBalance,
+        );
+    } else {
+        /**
+         * Overall logic to build sequence of single token exit calls:
+         * 1. Go BOTTOM-UP building exit path to tokenOut
+         * 2. Go through exit path filling out input amounts and output refs
+         * 3. Inputs will be bptAmountIn provided or output of the previous level
+         * 4. Output at bottom level is the amountOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+        const topPool = poolsTopDown[0];
+
+        // Go BOTTOM-UP building exit path to tokenOut
+        const exitPath: NestedPool[] = [];
+        let tokenOutByLevel = tokenOut;
+        while (tokenOutByLevel !== topPool.address) {
```

Comment:
Could probably be split into a separate a separate function with a description of whats happening.